watcher/handlers/metalogger.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package handlers contains handlers that implements Kubernetes toolscache.ResourceEventHandler.
+// The hankders are expected to be used with watchers and handle events dispatched by watchers.
+package handlers
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	mlpb "ml_metadata/proto/metadata_store_go_proto"
+	storepb "ml_metadata/proto/metadata_store_service_go_proto"
+
+	"github.com/golang/protobuf/proto"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const workspace = "resource_watcher"
+
+// MetaLogger logs newly created k8s resources into metadata service.
+type MetaLogger struct {
+	// Metadata gRPC client.
+	kfmdClient storepb.MetadataStoreServiceClient
+	// GroupVerionKind of the resource being watched.
+	resource schema.GroupVersionKind
+	typeID   int64
+}
+
+// NewMetaLogger creates a new MetaLogger for a specific k8s GroupVersionKind.
+func NewMetaLogger(kfmdClient storepb.MetadataStoreServiceClient, gvk schema.GroupVersionKind) (*MetaLogger, error) {
+	l := &MetaLogger{
+		resource:   gvk,
+		kfmdClient: kfmdClient,
+	}
+	resourceArtifactType := mlpb.ArtifactType{
+		Name: proto.String(l.MetadataArtifactType()),
+		Properties: map[string]mlpb.PropertyType{
+			// same as metav1.Object.Name
+			"name": mlpb.PropertyType_STRING,
+			// same as metav1.Object.UID
+			"version": mlpb.PropertyType_STRING,
+			// same as metav1.Object.Time in rfc3339 format
+			"create_time": mlpb.PropertyType_STRING,
+			// stores the metav1.Object
+			"object": mlpb.PropertyType_STRING,
+		},
+	}
+	request := storepb.PutArtifactTypeRequest{
+		ArtifactType:   &resourceArtifactType,
+		AllFieldsMatch: proto.Bool(true),
+	}
+	resp, err := kfmdClient.PutArtifactType(context.Background(), &request)
+	if err != nil {
+		return l, fmt.Errorf("failed to create artifact type: err = %v; request = %v; response = %v", err, request, resp)
+	}
+	l.typeID = resp.GetTypeId()
+	return l, nil
+}
+
+// MetadataArtifactType returns the metadata artifact type for the MetaLogger's GroupVerionKind
+func (l *MetaLogger) MetadataArtifactType() string {
+	return fmt.Sprintf("kubeflow.org/%s/%s", l.resource.GroupKind(), l.resource.Version)
```

Comment:
Are we only watching KF resources?

Reply:
No. The resource to be watched is configured by the `resourcelist` flag. By default, they are `pod` and `service`

this type is the metadata artifact type. I'm add a kubeflow prefix here to distinguish it from others.